test/extensions/clusters/aggregate/cluster_integration_test.cc

```diff
@@ -0,0 +1,189 @@
+  Http::ContextImpl http_context_;
+
+  const std::string default_yaml_config_ = R"EOF(
+ static_resources:
+  clusters:
+  - name: aggregate_cluster
+    connect_timeout: 0.25s
+    lb_policy: CLUSTER_PROVIDED
+    cluster_type:
+      name: envoy.clusters.aggregate
+      typed_config:
+        "@type": type.googleapis.com/envoy.config.cluster.aggregate.ClusterConfig
+        clusters:
+        - primary
+        - secondary
+  )EOF";
+};
+
+TEST_F(AggregateClusterIntegrationTest, NoHealthyUpstream) {
+  initialize(default_yaml_config_);
+  EXPECT_EQ(nullptr, cluster_->loadBalancer().chooseHost(nullptr));
+}
+
+TEST_F(AggregateClusterIntegrationTest, BasicFlow) {
+  initialize(default_yaml_config_);
+
+  std::unique_ptr<Upstream::MockClusterUpdateCallbacks> callbacks(
+      new NiceMock<Upstream::MockClusterUpdateCallbacks>());
+  Upstream::ClusterUpdateCallbacksHandlePtr cb =
+      cluster_manager_->addThreadLocalClusterUpdateCallbacks(*callbacks);
+
+  EXPECT_TRUE(cluster_manager_->addOrUpdateCluster(Upstream::defaultStaticCluster("primary"), ""));
+  auto primary = cluster_manager_->get("primary");
+  EXPECT_NE(nullptr, primary);
+  auto host = cluster_->loadBalancer().chooseHost(nullptr);
+  EXPECT_NE(nullptr, host);
+  EXPECT_EQ("primary", host->cluster().name());
+  EXPECT_EQ("127.0.0.1:11001", host->address()->asString());
+
+  EXPECT_TRUE(
+      cluster_manager_->addOrUpdateCluster(Upstream::defaultStaticCluster("secondary"), ""));
+  auto secondary = cluster_manager_->get("secondary");
+  EXPECT_NE(nullptr, secondary);
+  host = cluster_->loadBalancer().chooseHost(nullptr);
+  EXPECT_NE(nullptr, host);
+  EXPECT_EQ("primary", host->cluster().name());
+  EXPECT_EQ("127.0.0.1:11001", host->address()->asString());
+
+  EXPECT_TRUE(cluster_manager_->addOrUpdateCluster(Upstream::defaultStaticCluster("tertiary"), ""));
+  auto tertiary = cluster_manager_->get("tertiary");
+  EXPECT_NE(nullptr, tertiary);
+  host = cluster_->loadBalancer().chooseHost(nullptr);
+  EXPECT_NE(nullptr, host);
+  EXPECT_EQ("primary", host->cluster().name());
+  EXPECT_EQ("127.0.0.1:11001", host->address()->asString());
+
+  EXPECT_TRUE(cluster_manager_->removeCluster("primary"));
+  EXPECT_EQ(nullptr, cluster_manager_->get("primary"));
+  host = cluster_->loadBalancer().chooseHost(nullptr);
+  EXPECT_NE(nullptr, host);
+  EXPECT_EQ("secondary", host->cluster().name());
+  EXPECT_EQ("127.0.0.1:11001", host->address()->asString());
+  EXPECT_EQ(3, cluster_manager_->activeClusters().size());
+
+  EXPECT_TRUE(cluster_manager_->addOrUpdateCluster(Upstream::defaultStaticCluster("primary"), ""));
+  primary = cluster_manager_->get("primary");
+  EXPECT_NE(nullptr, primary);
+  host = cluster_->loadBalancer().chooseHost(nullptr);
+  EXPECT_NE(nullptr, host);
+  EXPECT_EQ("primary", host->cluster().name());
+  EXPECT_EQ("127.0.0.1:11001", host->address()->asString());
+
+  // Set up the HostSet with 1 healthy, 1 degraded and 1 unhealthy.
```

Comment:
id probably split this section out into its own test case, this test is pretty long